packages/core/src/StateNode.ts

```diff
@@ -1087,17 +1087,17 @@ class StateNode<
     const exitActions = Array.from(exitStates).map((stateNode) =>
       toActionObjects(
         [
-          ...stateNode.onExit,
-          ...stateNode.activities.map((activity) => stop(activity))
+          ...stateNode.activities.map((activity) => stop(activity)),
+          ...stateNode.onExit
         ],
         this.machine.options.actions as any
       )
     );
 
-    const actions = exitActions
-      .concat([
-        toActionObjects(transition.actions, this.machine.options.actions as any)
-      ])
+    const actions = [
+      toActionObjects(transition.actions, this.machine.options.actions as any)
+    ]
+      .concat(exitActions)
```

Comment:
This one is definitely a wrong change. All exit actions should be executed before transition actions and entry actions should be executed after transition actions:
1. exit actions
2. transition actions
3. entry actions

https://www.w3.org/TR/scxml/#microstepProcedure